system/lib/wasmfs/syscalls.cpp

```diff
@@ -719,4 +719,130 @@ long __syscall_getdents64(long fd, long dirp, long count) {
+  // other threads could unlink the destination directory after we have moved
+  // our source child. This would leave the file system in an undefined state as
+  // the source child is now orphaned. Protecting the destination and source
+  // parent directories ensures that no resources that we depend on are altered.
+  // Trylocking on the new_path parent is needed in the case where two renames
+  // are operating on opposing sources and destinations. This would cause them
+  // to lock the directories in reverse order and could cause deadlock.
+
+  // Edge case: rename("dir", "dir/somename") - in this scenario it should not
+  // be possible to rename the destination if the source is an ancestor.
+
+  // Edge case: rename("dir/somename", "dir") - in this scenario it should not
+  // be possible to rename the destination since by definition it is non-empty.
+
+  auto oldPathParts = splitPath((char*)old_path);
+
+  if (oldPathParts.empty()) {
+    return -EINVAL;
+  }
+
+  // In Linux, renaming the root directory returns EBUSY.
+  // TODO: Fix this when path parsing is refactored.
+  if (oldPathParts.size() == 1 && oldPathParts[0] == "/") {
+    return -EBUSY;
+  }
+
+  auto oldBase = oldPathParts.back();
+
+  long err;
+  auto oldParentDir = getDir(oldPathParts.begin(), oldPathParts.end() - 1, err);
+
+  if (!oldParentDir) {
+    return err;
+  }
+
+  // For this operation to be atomic we must lock the source parent directory.
+  auto lockedOldParentDir = oldParentDir->locked();
+
+  auto oldPath = lockedOldParentDir.getEntry(oldBase);
+
+  if (!oldPath) {
+    return -ENOENT;
+  }
+
+  // Obtain the destination parent directory to see if it exists.
+  auto newPathParts = splitPath((char*)new_path);
+
+  if (newPathParts.empty()) {
+    return -EINVAL;
+  }
+
+  // In Linux, renaming a directory with the root directory returns ENOTEMPTY.
+  // TODO: Fix this when path parsing is refactored.
+  if (newPathParts.size() == 1 && newPathParts[0] == "/") {
+    return -ENOTEMPTY;
+  }
+
+  auto newBase = newPathParts.back();
+
+  auto newParentDir =
+    getDir(newPathParts.begin(), newPathParts.end() - 1, err, oldPath);
+
+  // If the destination parent directory doesn't exist, the source file cannot
+  // be moved.
+  if (!newParentDir) {
+    return err;
+  }
+
+  // Edge case: a/b -> a/c share the same parent. Trylocking the same parent
+  // with a recursive mutex should work, so we don't need to treat that case
+  // specially here.
+  auto maybeLockedNewParentDir = newParentDir->maybeLocked();
+  if (!maybeLockedNewParentDir) {
+    return -EBUSY;
+  }
+
+  auto lockedNewParentDir = std::move(*maybeLockedNewParentDir);
+  auto newPath = lockedNewParentDir.getEntry(newBase);
+
+  // If old_path and new_path are the same, do nothing.
+  if (newPath == oldPath) {
+    return 0;
+  }
+
+  // Cannot move to a destination parent directory wihout write permissions.
+  if (!(lockedNewParentDir.mode() & WASMFS_PERM_WRITE)) {
+    return -EACCES;
+  }
```

Comment:
Don't we need the same permission on the old path? We are modifying that directory as well.

Reply:
Hmm good point, this is a good case to add to the tests as well.